### ISSUE TITLE
Cloud Storage support in the file browser

### ISSUE THREAD

**Issue by Zorlin** [OSC/ondemand#1789]
Hi there,

We are exploring Open OnDemand at [my workplace](https://github.com/PawseySC), but would like to use it with an S3-based object storage system.

Is there a possibility of getting S3 support added to the files app?

[@johrstrom's edits] -- Yes! But it'll likely need to be community driven as OSC doesn't use cloud storage systems.

* [x] #1828
* [x] #1904
* [x] #2119
* [x] #2129
* [x] #2231
* [x] #2232
* [x] #2247

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202815755532406/1202822052279939) by [Unito](https://www.unito.io)

**Comment by treydock** [comment]
You could mount the S3 bucket with something like S3FS and then it would be available to OnDemand if the permissions allow the user logged into OnDemand to access the files.

**Comment by Zorlin** [comment]
That's a good idea but would require mounting each container as needed for each user. :)

**Comment by johrstrom** [comment]
Yea it seems like we could add a button that mounts a bucket for the user. 

That's an easy entry-point because we can then treat it like the existing file system and don't really have to update the UI a lot.

**Comment by johrstrom** [comment]
@Zorlin I'm going to start to update your first comment - I hope that's OK, but it's in the interest of moving this forward.


@robinkar you're interested in this feature yes?  https://edgeguides.rubyonrails.org/active_storage_overview.html looks like it directly supports this so I'd say that's route we'd be heading.  I'm going to start to use this ticket to track other tickets to sort of map the work I think it would take to do this.

**Comment by johrstrom** [comment]
Sorry - I actually need to duplicate this ticket so that i can edit the initial comment for tracking purposes.

To answer the initial question of will we support it - I'd say it's very likely given `active_storage`'s support for it, but we'd likely need some contribution as we don't use cloud storage.

**Comment by johrstrom** [comment]
Sorry again for the back and forth. Let's use this ticket for tracking because honestly we don't get a lot of community requests, so let's cherish them.

**Comment by robinkar** [comment]
I have to take a closer look at Active Storage, but this is still a feature we are looking for. I am currently preparing a small patch to bring us a bit closer to this goal.

**Comment by johrstrom** [comment]
Cool. I just mention it because I'm sort of dependency adverse, but since it's a part of rails that makes it like 1000% more attractive.

**Comment by gerald-byrket** [comment]
Hi @Zorlin,

Is this something that you will be able to complete for our 2.1 release?  If so, I will add it to our 2.1 Release Project.

Thanks,
-gerald

**Comment by Zorlin** [comment]
Hi @gerald-byrket - I won't personally be able to develop the feature (much), but would like to strongly advocate for it. Our centre is all in on object storage and I can't imagine we're the only one :)

**Comment by gerald-byrket** [comment]
Bummer.

I was hoping you would own the implementation.  Ok.  We’ll have to prioritize based on us having to do it.

Thanks,
Gerald Byrket
***@***.******@***.***>
817-507-6993

**Comment by robinkar** [comment]
I've had some more time these last few weeks to think about this, and made some progress. I'll see if I could get some more PRs submitted soon with some changes.
It would be good to get a decision on what the implementation would be based on (Rclone, Active Storage, etc.). Currently I have a working prototype with file copying, moving, renaming, deleting, viewing and downloading using Rclone, as that works surprisingly well (though there's a few problematic cases that will have to be solved). If that's the route we would like to go I could probably submit some other parts of the code to get the implementation started.

**Comment by johrstrom** [comment]
I'm happy to use Rclone especially if active storage takes a lot of assumptions. Rclone has lots of back end support and is directly installable through an RPM or deb package (and is optional to boot) so that's all good really.

**Comment by Zorlin** [comment]
Ah! This is awesome :) I had a quick go of it and was unable to really get it working properly - visiting /minio_s3:/ didn't work (via Change Directory) but if I navigated manually to $server/pun/sys/dashboard/files/minio_s3/ I could see my buckets, browse them and even download files.

Uploading files didn't work, nor did deleting, renaming and a few other things I tried. But it's a huge start and I'm grateful it's working even this much!

Any tips on getting it working further?

**Comment by Zorlin** [comment]
<img width="538" alt="Screen Shot 2022-08-29 at 6 17 24 pm" src="https://user-images.githubusercontent.com/1369772/187179444-1ea19717-50bb-4cfd-ad3e-81b0a82c6436.png">

**Comment by robinkar** [comment]
> Ah! This is awesome :) I had a quick go of it and was unable to really get it working properly - visiting /minio_s3:/ didn't work (via Change Directory) but if I navigated manually to $server/pun/sys/dashboard/files/minio_s3/ I could see my buckets, browse them and even download files.
> 
> Uploading files didn't work, nor did deleting, renaming and a few other things I tried. But it's a huge start and I'm grateful it's working even this much!
> 
> Any tips on getting it working further?

The *Change Directory* button only navigates within the current filesystem/remote, but this should probably be changed to also allow navigating between them. #2128 is another idea for the navigation could work. Currently, the best way to handle the navigation is to setup favorite paths for the remotes.

Is there anything interesting in the dashboard logs or browser console for the actions that don't work? My first guess would be that the dashboard assets (javascript) need to be rebuilt.

**Comment by Zorlin** [comment]
Dumb question, but where are the logs to tell? I've been trawling through
the documentation and haven't found where logs are :)

**Comment by robinkar** [comment]
> Dumb question, but where are the logs to tell? I've been trawling through the documentation and haven't found where logs are :)

I believe the error logs should be at `/var/log/ondemand-nginx/$USER/error.log`.

**Comment by robinkar** [comment]
@Zorlin Does the normal local filesystem behave as expected? I'm running into #1193 when running the dashboard as `/pun/sys/dashboard` in the dev container so both remote and local uploads, renames etc. don't work without the workaround there.

**Comment by Zorlin** [comment]
Hi @robinkar, yes, normal filesystem works fine. We're using HTTPS already, which seems to be the cause of that issue (if I'm reading that thread correctly).

Here are the logs from attempting to upload a file. This is after running `rclone config` as my user.
```
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "execve = [\"rclone\", \"lsf\", \"--low-level-retries=1\", \"minio_s3:/\"]"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3/ood-test format=html controller=FilesController action=fs status=200 duration=282.92 view=17.39"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "execve = [\"rclone\", \"lsf\", \"--low-level-retries=1\", \"minio_s3:/\"]"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "execve = [\"rclone\", \"lsjson\", \"--low-level-retries=1\", \"minio_s3:/ood-test\"]"
App 6847 output: [2022-08-30 10:39:44 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3/ood-test format=html controller=FilesController action=fs status=200 duration=386.70 view=6.19"
App 6847 output: [2022-08-30 10:39:50 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:39:50 +0800 ]  INFO "execve = [\"rclone\", \"moveto\", \"/tmp/RackMultipart20220830-6847-oksyml.png\", \"minio_s3:/ood-test/ansible-black.png\"]"
App 6847 output: [2022-08-30 10:39:51 +0800 ]  INFO "method=POST path=/pun/sys/dashboard/files/upload/minio_s3 format=*/* controller=FilesController action=upload status=500 duration=209.87 view=0.81"
```

This is after moving `.config/rclone` to `.config/rclone-disabled`:
```
App 6847 output: [2022-08-30 10:42:46 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "execve = [\"rclone\", \"lsf\", \"--low-level-retries=1\", \"minio_s3:/\"]"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3/ood-test format=html controller=FilesController action=fs status=200 duration=269.33 view=27.92"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "execve = [\"rclone\", \"lsf\", \"--low-level-retries=1\", \"minio_s3:/\"]"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "execve = [\"rclone\", \"lsjson\", \"--low-level-retries=1\", \"minio_s3:/ood-test\"]"
App 6847 output: [2022-08-30 10:42:47 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3/ood-test format=html controller=FilesController action=fs status=200 duration=366.10 view=7.15"
App 6847 output: [2022-08-30 10:42:55 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:42:55 +0800 ]  INFO "execve = [\"rclone\", \"moveto\", \"/tmp/RackMultipart20220830-6847-kamrkc.png\", \"minio_s3:/ood-test/ansible-black.png\"]"
App 6847 output: [2022-08-30 10:42:55 +0800 ]  INFO "method=POST path=/pun/sys/dashboard/files/upload/minio_s3 format=*/* controller=FilesController action=upload status=500 duration=178.30 view=0.67"
```

Loading and listing files works fine, as does downloading them. Actions that currently don't work - 

* editing
* renaming
* deleting
* copying/moving

Copying a file to a different bucket results in the following logs:
```
App 6847 output: [2022-08-30 10:44:31 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3 format=html controller=FilesController action=fs status=200 duration=207.92 view=9.47"
App 6847 output: [2022-08-30 10:44:33 +0800 ]  INFO "execve = [\"rclone\", \"listremotes\", \"--long\"]"
App 6847 output: [2022-08-30 10:44:33 +0800 ]  INFO "execve = [\"rclone\", \"lsf\", \"--low-level-retries=1\", \"minio_s3:/\"]"
App 6847 output: [2022-08-30 10:44:33 +0800 ]  INFO "execve = [\"rclone\", \"lsjson\", \"--low-level-retries=1\", \"minio_s3:/koala-backups\"]"
App 6847 output: [2022-08-30 10:44:33 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/files/minio_s3/koala-backups format=html controller=FilesController action=fs status=200 duration=303.03 view=10.30"
App 6847 output: [2022-08-30 10:44:34 +0800 ]  INFO "method=POST path=/pun/sys/dashboard/transfers.json format=json controller=TransfersController action=create status=500 error='Errno::ENOENT: No such file or directory @ rb_file_s_lstat - /ood-test' duration=44.52 view=0.00"
App 6847 output: [2022-08-30 10:44:34 +0800 ] FATAL "Errno::ENOENT (No such file or directory @ rb_file_s_lstat - /ood-test):\n  \napp/models/posix_file.rb:16:in `lstat'\napp/models/posix_file.rb:16:in `lstat'\napp/models/posix_file.rb:16:in `rescue in stat'\napp/models/posix_file.rb:13:in `stat'\napp/models/local_transfer.rb:127:in `mv_to_same_device?'\napp/models/local_transfer.rb:123:in `synchronous?'\napp/controllers/transfers_controller.rb:24:in `create'"
App 6847 output: [2022-08-30 10:44:34 +0800 ]  INFO "method=GET path=/pun/sys/dashboard/500 format=json controller=ErrorsController action=internal_server_error status=500 error='ActionView::MissingTemplate: Missing template errors/internal_server_error, application/internal_server_error with {:locale=>[:en], :formats=>[:json], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :md, :markdown, :jbuilder]}. Searched in:\n  * \"/var/www/ood/apps/sys/dashboard/app/views\"\n  * \"/opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/ood_appkit-2.1.1/app/views\"\n' duration=8.39 view=0.00"
App 6847 output: Error during failsafe response: Missing template errors/internal_server_error, application/internal_server_error with {:locale=>[:en], :formats=>[:json], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :md, :markdown, :jbuilder]}. Searched in:
App 6847 output:   * "/var/www/ood/apps/sys/dashboard/app/views"
App 6847 output:   * "/opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/ood_appkit-2.1.1/app/views"
App 6847 output:
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/path_set.rb:48:in `find'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/lookup_context.rb:127:in `find'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/renderer/template_renderer.rb:45:in `determine_template'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/renderer/template_renderer.rb:7:in `render'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/renderer/renderer.rb:61:in `render_template_to_object'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/renderer/renderer.rb:29:in `render_to_object'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/rendering.rb:117:in `block in _render_template'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/base.rb:273:in `in_rendering_context'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/rendering.rb:116:in `_render_template'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/streaming.rb:218:in `_render_template'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/rendering.rb:103:in `render_to_body'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/rendering.rb:52:in `render_to_body'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/renderers.rb:142:in `render_to_body'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/abstract_controller/rendering.rb:25:in `render'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/rendering.rb:36:in `render'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:46:in `block (2 levels) in render'
App 6847 output:   /opt/rh/rh-ruby30/root/usr/share/ruby/benchmark.rb:308:in `realtime'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/core_ext/benchmark.rb:14:in `ms'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:46:in `block in render'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:86:in `cleanup_view_runtime'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:45:in `render'
App 6847 output:   /var/www/ood/apps/sys/dashboard/app/controllers/errors_controller.rb:31:in `block (2 levels) in internal_server_error'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/mime_responds.rb:214:in `respond_to'
App 6847 output:   /var/www/ood/apps/sys/dashboard/app/controllers/errors_controller.rb:29:in `internal_server_error'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/abstract_controller/base.rb:228:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/rendering.rb:30:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/abstract_controller/callbacks.rb:42:in `block in process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/callbacks.rb:106:in `run_callbacks'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/abstract_controller/callbacks.rb:41:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/rescue.rb:22:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/notifications.rb:203:in `block in instrument'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/notifications.rb:203:in `instrument'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb:33:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal/params_wrapper.rb:249:in `process_action'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/abstract_controller/base.rb:165:in `process'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionview-6.1.5.1/lib/action_view/rendering.rb:39:in `process'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal.rb:190:in `dispatch'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_controller/metal.rb:254:in `dispatch'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb:33:in `serve'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb:50:in `block in serve'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb:32:in `each'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb:32:in `serve'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb:842:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/show_exceptions.rb:53:in `render_exception'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/show_exceptions.rb:36:in `rescue in call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/lograge-0.12.0/lib/lograge/rails_ext/rack/logger.rb:18:in `call_app'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/railties-6.1.5.1/lib/rails/rack/logger.rb:26:in `block in call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib
App 6847 output: /active_support/tagged_logging.rb:99:in `block in tagged'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/tagged_logging.rb:37:in `tagged'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/tagged_logging.rb:99:in `tagged'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/railties-6.1.5.1/lib/rails/rack/logger.rb:26:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/request_store-1.5.1/lib/request_store/middleware.rb:19:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/request_id.rb:26:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/rack-2.2.3.1/lib/rack/method_override.rb:24:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/rack-2.2.3.1/lib/rack/runtime.rb:22:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/activesupport-6.1.5.1/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/executor.rb:14:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/rack-2.2.3.1/lib/rack/sendfile.rb:110:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/host_authorization.rb:148:in `call'
App 6847 output:   /opt/ood/ondemand/root/usr/share/gems/3.0/ondemand/2.1.20220828/gems/railties-6.1.5.1/lib/rails/engine.rb:539:in `call'
App 6847 output:   /opt/rh/ondemand/root/usr/share/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:107:in `process_request'
App 6847 output:   /opt/rh/ondemand/root/usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:157:in `accept_and_process_next_request'
App 6847 output:   /opt/rh/ondemand/root/usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:110:in `main_loop'
App 6847 output:   /opt/rh/ondemand/root/usr/share/ruby/vendor_ruby/phusion_passenger/request_handler.rb:419:in `block (3 levels) in start_threads'
App 6847 output:   /opt/rh/ondemand/root/usr/share/ruby/vendor_ruby/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'
```

Happy to kick the tires on this until it's solid, we're very keen to see it ship! :)

**Comment by Zorlin** [comment]
FWIW: taking a file and copying it into a bucket with rclone manually works fine:
`rclone copyto ~/pawsey-logo-beige-32px.png minio_s3:/ood-test/pawsey-logo-beige-32px.png`

**Comment by robinkar** [comment]
Are you using the nightly rpms? If I'm understanding everything correctly, those are still on the `a35d485` commit, which means they are still missing the implementation for the transfers.

**Comment by Zorlin** [comment]
Yes, I am. That would probably explain what I'm seeing :)

**Comment by Zorlin** [comment]
Might try installing OOD from the latest source if that's viable, so I can provide feedback on the more current tree.

**Comment by Zorlin** [comment]
Yep, now on latest Git version, built from source (OnDemand version: 440b222), and getting the same results.

**Comment by robinkar** [comment]
> Yep, now on latest Git version, built from source (OnDemand version: [440b222](https://github.com/OSC/ondemand/commit/440b2228baf9cf7d32ac934d10ff9e17f566e237)), and getting the same results.

That's really strange. Did file editing start working by any chance?

If you check the `POST /pun/sys/dashboard/transfers.json` request in the browser console when moving files, what does it look like? It should look something like this:
```
{
  "command": "mv",
  "files": {
    "/mybucket/myfile": "/mybucket/renamedfile"
  },
  "from_fs": "minio_s3",
  "to_fs": "minio_s3"
}

```
If those `from_fs` and `to_fs` parameters are missing, ensure you have run `bin/setup` for the dashboard.

**Comment by Zorlin** [comment]
Editing is closer to working than it was before - it successfully opens the file - but saving the file doesn't actually save it, despite the jQuery query it triggers returning a 200 OK.

**Comment by Zorlin** [comment]
This is the transfers.json I collected:

```
{"command":"cp","files":{"/ood-test/NanoNano.sh":"/koala-backups/NanoNano.sh"},"from_fs":"minio_s3","to_fs":"minio_s3"}
```

**Comment by Zorlin** [comment]
Changing 
`RCLONE_CONFIG_MINIO_S3_ACL: "PRIVATE"`
to `RCLONE_CONFIG_MINIO_S3_ACL: "private"`

and
`RCLONE_CONFIG_MINIO_S3_PROVIDER: "Other"`
to `RCLONE_CONFIG_MINIO_S3_PROVIDER: "Ceph"`

seems to have fixed the issue. I can now copy, edit, move, delete, rename files successfully.

**Comment by robinkar** [comment]
I thought I should probably let you guys know that I will be starting exchange studies soon and won't be working at CSC, so someone else from CSC (most likely) will be contributing towards this feature instead of me.

**Comment by Zorlin** [comment]
@robinkar A big and sincere thank you for making the feature happen. The functionality is there now, I think the remaining work is largely UX. I wish you well in your future study and work.

**Comment by johrstrom** [comment]
@robinkar let me second that big and sincere thank you!

**Comment by johrstrom** [comment]
This is basically complete and will be released 2.1.
